Add anchor for section titles for quick sharing of sections

marinaaisa opened the pull request:
Bug/issue #96943502, if applicable: 

## Summary

Add anchor for section titles for quick sharing of sections.
This anchor shows when user hovers the title section.

Issue: https://github.com/apple/swift-docc-render/issues/273

## Dependencies

NA

## Testing

Steps:
1. Open any docc archive
2. Assert that all h2 and h3 titles in documentation pages show an anchor when hovering the title

## Checklist

Make sure you check off the following items. If they cannot be completed, provide a reason.

- [x] Added tests
- [x] Ran `npm test`, and it succeeded
- [x] Updated documentation if necessary

marinaaisa commented:
@swift-ci test

marinaaisa commented:
@swift-ci test

dobromir-hristov commented on src/components/ContentNode/SectionTitle.vue:
```diff
@@ -0,0 +1,75 @@
+
+<template>
+  <component
+   :id="anchor ? anchor : null"
```
```suggestion
   :id="anchor || null"
```

dobromir-hristov commented on src/components/ContentNode/SectionTitle.vue:
```diff
@@ -0,0 +1,75 @@
+  >
+    <a
+      v-if="anchor || href"
+      :href="`#${anchor || href}`"
```
I would suggest you look at how we have done this in the past, like in the TutorialsNavigationLink component.

We use a `scrollToElement` function that actually adjusts the scrolling by a few pixels, so the navigation does not overlap the title. I saw you added `scroll-margin-top` but I am getting cutoffs on the Topic section

We also used to use `<router-link>` components instead of `a`, not sure if there is a major diff.

![image](https://user-images.githubusercontent.com/9863944/184084450-a4774561-20a3-4c8e-9a64-2bcebee839e9.png)

dobromir-hristov replied:
So if we pass `anchor`, it adds an ID, and if we pass just `href` it does not, and it uses both to create a `href`?

I had to go back midway review when I finally got whats going on, felt confusing. I will leave it the the rest of the team to give more feedback.

marinaaisa replied:
I addressed the feedback and I fixed the confusion with the anchor and href prop. Now we only have one. Check it out again for another review, thanks!

dobromir-hristov commented on src/components/DocumentationTopic/ContentTableSection.vue:
```diff
@@ -24,14 +24,20 @@
     },
   },
+  computed: {
+    anchor: ({ title }) => title.replaceAll(' ', '-'),
```
we have an `anchorize` function, use that instead.

dobromir-hristov reviewed (commented):
This looks great, very nice! 🚀 

One thing I noticed is if I tab through, it does not show the anchors, deff something we should fix.

Need to check if we should disable this for IDE environments.

Also we may check with the DocC team if they plan to provide anchors for those Topic subsections, instead of us generating them - #190

Other than that, looks great :)

marinaaisa commented:
Thanks @dobromir-hristov !

> One thing I noticed is if I tab through, it does not show the anchors, deff something we should fix.

Fixed! Good catch!

> Need to check if we should disable this for IDE environments.

Why would we? it should scroll in the same way, right?
 
> Also we may check with the DocC team if they plan to provide anchors for those Topic subsections, instead of us generating them - #190

Yeah, that would be much better than the generation I'm doing in the frontend!

Give it another review. I still need to add final tests for the `handleFocusAndScroll` function I added in the scrollToElement mixin.

dobromir-hristov commented:
I don't see a point in having a clickable link in the ide, since there is no URL bar

dobromir-hristov reviewed (commented):
I think it looks good now. How does VO handle the hidden anchor item? Do we need anything special for it?

marinaaisa commented:
> I think it looks good now. How does VO handle the hidden anchor item? Do we need anything special for it?

We are hiding it from VO because I think it's not providing any value to VO users. https://vuejs.org/guide/introduction.html#what-is-vue does it in the same way

marinaaisa commented on src/components/ContentNode/SectionTitle.vue:
```diff
@@ -10,51 +10,57 @@
   },
+  mounted() {
+    if (!this.anchor) return;
+    const element = document.getElementById(`${this.anchor}`);
```
Unpredictable? We use dynamic ids in many more components across the app... 🤔 
But if you have any better approach, happy to look into it :)

dobromir-hristov replied:
Well you may be right. I would have probably just made this logic user controllable, via some boolean whether to apply an `id` to the element or not.

marinaaisa commented:
> I don't see a point in having a clickable link in the ide, since there is no URL bar

True, I don't render it anymore on IDEs https://github.com/apple/swift-docc-render/pull/408/commits/f6bd656caaab0e1c853e5f416cdceca93f6d273c

marinaaisa commented:
@swift-ci test

dobromir-hristov reviewed (commented):
Looks good. @mportiz08 can you look into the anchor logic if you are fine with it and we can merge this one I think

mportiz08 commented on src/components/ContentNode.vue:
```diff
@@ -24,6 +25,7 @@ import StrikeThrough from './ContentNode/StrikeThrough.vue';
 const BlockType = {
   aside: 'aside',
   codeListing: 'codeListing',
+  sectionTitle: 'sectionTitle',
```
I don't think this line should needed since these enum values refer to content node block types from the Render Node spec (and this is not one).

mportiz08 commented on src/components/ContentNode.vue:
```diff
@@ -227,14 +229,13 @@ function renderNode(createElement, references) {
+    case BlockType.heading: {
+      const props = {
+        anchor: node.anchor,
+        tag: `h${node.level}`,
```
I think maybe we should pass the `level` value as a prop and then map those to the right header tag in the new component—otherwise we allow for any arbitrary string to be used as the HTML tag for this component when we probably only want to allow h1-h6.

marinaaisa replied:
done!

mportiz08 commented on src/components/ContentNode/SectionTitle.vue:
```diff
@@ -0,0 +1,86 @@
+import scrollToElement from 'docc-render/mixins/scrollToElement';
+
+export default {
+  name: 'SectionTitle',
```
I think we might want to rename this to better indicate that this component is used for headings. Maybe just `Heading` or `LinkableHeading`? That would also more closely align with the node kind from the Render Node spec (not necessary but maybe nice).

marinaaisa replied:
I liked it! I renamed it to `LinkableHeading` :)

mportiz08 commented on src/components/ContentNode/SectionTitle.vue:
```diff
@@ -0,0 +1,86 @@
+  },
+  mounted() {
+    if (!this.anchor) return;
+    const element = document.getElementById(`${this.anchor}`);
```
```suggestion
    const element = document.getElementById(this.anchor);
```

As a more general comment though, I don't totally understand the need for why this hook is needed. Why are there separate `id` and `anchor` values and how are they used differently?

dobromir-hristov replied:
There are cases where the ID is applied on a section that wraps the heading (like in TopicsSection), but in other cases, the heading has to hold the ID.

mportiz08 replied:
I'm not totally following how that requires the mounted hook here (probably just missing something).

dobromir-hristov replied:
I think she wanted it to be more magical. Thats one of the points I raised that I wanted your input on :D 

The component tries to find an already existing element with that `id` in the DOM, and if it doesnt, it adds that `id` to itself.

mportiz08 replied:
I guess I just don't understand the magical part, to be more clear. Why would this logic have to happen after the initial render?

dobromir-hristov replied:
So all possible items that may have that ID are rendered? As I said I would probably control this through a prop instead.

marinaaisa replied:
@mportiz08 I had this logic to check if the document has already an `id` on it where we want to link to. This happens in components that have the id in the beginning of the section itself instead of having it in the title.
Also in this way, we avoid duplication of `id`s in the HTML, since ID attributes must be unique in the document. 

Controlling through a prop, as @dobromir-hristov suggested, would allow us more flexibility but would open the component for possible HTML's errors because we wouldn't prevent the document from duplication of HTML `id`s.

mportiz08 replied:
I understand why we might choose to sometimes assign the ID attribute to a non-heading element instead of a heading, but I don't understand why we couldn't prevent the ID from being assigned in multiple places in the components themselves—it seems like that is controlled by our app and not outside forces. If that's the case, then we should be able to programmatically decide where the anchor uniquely goes before any `mounted` hook gets called after the initial render.

I think we both agree on the goal of only having unique IDs but maybe there's a little confusion about how to ensure it within the components themselves. Let me know if you want to talk about this offline—I may be missing a subtle thing here that might make this approach necessary.

marinaaisa replied:
@mportiz08 I set a meeting to continue discussing this. Interesting debate! Thanks :)

mportiz08 commented on src/components/ContentNode/SectionTitle.vue:
```diff
@@ -0,0 +1,86 @@
+
+<template>
+  <component
+   :id="!isTargetIDE ? id : null"
```
Can this just be `anchor` or am I missing something?

Also, I don't think we need to strip out the `id` attribute even if the actual link to that anchor isn't presented in certain situations (like in IDE views).

dobromir-hristov replied:
We should strip out the UI element on hover/focus, but the ID here is fine to stay.

marinaaisa replied:
I left the id without changing it

mportiz08 commented on src/components/ContentNode/SectionTitle.vue:
```diff
@@ -0,0 +1,86 @@
+    },
+    tag: {
+      type: String,
+      default: () => 'h2',
```
Whether this stays as a `tag` or `level` prop, we should probably add a validator to ensure that you can only ever use a h1-h6 (or only a 1-6 level).

marinaaisa replied:
Added!

mportiz08 commented on src/components/ContentNode/SectionTitle.vue:
```diff
@@ -0,0 +1,86 @@
+.header-anchor {
+  margin-left: -0.73em;
+  padding-right: 0.23em;
+  transition: opacity .25s;
```
You may want to have someone from the design team review this new feature and test the experience in different browsers. The scrolling animation specifically (not this one) feels a little different in Safari but that may be a limitation of the JS scroll function itself.

mportiz08 commented on src/components/DocumentationTopic/ContentTable.vue:
```diff
@@ -15,18 +15,19 @@
   >
     <div class="container">
-      <h2 class="title">{{title}}</h2>
+      <SectionTitle class="title" :anchor="anchor">{{ title }}</SectionTitle>
```
We might need to do some additional investigation on this, but I suspect that we are now adding multiple elements with the same `id` value to the page since the `OnThisPageSection` is also using this `anchor` as an `id` for another HTML element, and there should only be one on the page since these are meant to be unique on a given page.

It might be as simple as removing it from the other component since it was previously used for the old "On This Page" affordance, which was removed with the introduction of the sidebar.

dobromir-hristov replied:
#416 :) Not quite :D

mportiz08 replied:
I know that component is coming back in a forthcoming PR, but we will need to ensure that we are only assigning a given unique ID to a single HTML element and not adding it in multiple places, regardless of whether or not there is an "On This Page" component being utilized.

dobromir-hristov replied:
Right, I think you can check my explanation on how this component works and that might make things more clear. 
As long as we dont touch the OnThisPage registration logic, I am good 🤣

marinaaisa replied:
Current logic should be enough to prevent duplication: https://github.com/apple/swift-docc-render/pull/408#discussion_r958671127

mportiz08 commented on src/components/DocumentationTopic/ContentTableSection.vue:
```diff
@@ -24,14 +24,21 @@
     },
   },
+  computed: {
+    anchor: ({ title }) => anchorize(title),
```
I know that DocC is not currently providing `anchor` values for these headings, but I think that work is being tracked since we ideally want it to be in control of the id as opposed to having it being generated it here—I think we could utilize `anchorize` as a fallback in the meantime, but we may run into future issues where old links might not work as one system takes over.

marinaaisa replied:
I added most of the logic from https://github.com/apple/swift-docc-render/pull/190 into this PR to address this

mportiz08 replied:
Good idea, thanks. There could be future issues when the switch happens where the anchors currently being generated in the app may conflict with futures ones provided by DocC, but I think that on-page link URLs are generally pretty unstable with titles changing anyways, so it probably makes sense to rely on DocC and only generate them when they aren't for now.

mportiz08 commented on src/mixins/scrollToElement.js:
```diff
@@ -27,5 +27,19 @@ export default {
+     * Always scroll to the element and focus it.
+     * This ensures that the next tab target is inside
+     * and that it is in view, even if the hash is in the url
+     * @returns {Promise<void>}
```
```suggestion
     * @returns {Promise<Element>}
```

marinaaisa replied:
fixed!

mportiz08 reviewed (commented):
Very cool to see this in action, and I love that it just works as expected in places where no anchor data is provided.

I have a minor concern about the anchor generation, but I think the majority of my other comments are pretty small things that will be easy to fix.

mportiz08 commented:
Just noting that we may want to close #190 when this gets merged.

dobromir-hristov commented:
> Just noting that we may want to close https://github.com/apple/swift-docc-render/pull/190 when this gets merged.

Even though this PR adds most of what #190 does, that one was assuming the anchors come from DocC. I think if Marina does that too, we can safely drop it :)

marinaaisa commented:
> Even though this PR adds most of what https://github.com/apple/swift-docc-render/pull/190 does, that one was assuming the anchors come from DocC. I think if Marina does that too, we can safely drop it :)

I moved most of the changes from https://github.com/apple/swift-docc-render/pull/190 PR into this one, so we can merge it in here.

marinaaisa commented:
@swift-ci test

mportiz08 commented on src/components/ContentNode/LinkableHeading.vue:
```diff
@@ -41,9 +43,10 @@ export default {
+    level: {
+      type: Number,
+      default: () => 2,
+      validator: v => v in HeadingLevels,
```
I don't think the `in` operator works like you're expecting it to here.

Example:
```javascript
> 0 in [1, 2, 3, 4, 5, 6]
true
> 6 in [1, 2, 3, 4, 5, 6]
false
```

It might be better to just check if the number is in bounds. Something like:
```javascript
v => v >= 1 && v <= 6
```

marinaaisa replied:
Oh, thank you! it seemed I was using it wrong, yeah :D fixed it!

mportiz08 commented on src/components/ContentNode.vue:
```diff
@@ -232,9 +231,9 @@ function renderNode(createElement, references) {
       const props = {
         anchor: node.anchor,
-        tag: `h${node.level}`,
+        level: Number(node.level),
```
I believe that `node.level` values are guaranteed to be integers in the RenderNode spec, so you can just pass `node.level` here without the `Number()` call.

marinaaisa replied:
fixed!

marinaaisa commented:
@swift-ci test

hqhhuang commented:
Hey Marina!
This is brilliant! I still have questions about the `id`, but I will reach out offline for that.
I tested this and noticed 3 small things:
1. The URL seems to be case sensitive. Is this intentional?
2. This issue appears in Chrome only:

https://user-images.githubusercontent.com/87735557/188030876-b5ce5a28-39f1-4cf0-b5dd-3ad611ae10f5.mov

3. Do we want to dynamically update the URL with the anchors based on where the user is on the page as they scroll? I think the Vue Doc site does this.

marinaaisa commented:
Hey @hqhhuang ! thanks a lot for your comments!

> 1. The URL seems to be case sensitive. Is this intentional?

URL are always case sensitive. You will see that some anchors are in lowercase and others aren't. This is because some come directly from docc (some have uppercase initials) and other we add them depending on the title (these ones are in lowercase). The anchor's names don't matter if they are in lowercase or not, as long as they point to an existing id with the same name. 

> 2. This issue appears in Chrome only:
>  Screen.Recording.2022-09-01.at.4.24.57.PM.mov

This is because of the way we scroll to the section. We first scroll into the view where the title is, and then we scrollBy to overcome the navigation height. You can see it here: https://github.com/apple/swift-docc-render/blob/f5c1b7fd271518503e85f7a5d2dd5f993ded17be/src/mixins/scrollToElement.js#L26-L27
I'm not sure yet how to prevent that jumping at the beginning, in theory, it shouldn't be noticeable... we could look into improving it in the future.

> 3. Do we want to dynamically update the URL with the anchors based on where the user is on the page as they scroll? I think the Vue Doc site does this.

Very good catch! Maybe this could a separate feature since it uses the ids but it doesn't depend on the new LinkableHeadinds, could you fill a radar or Github issue for it? Thanks!

mportiz08 commented:
I know you're still iterating on the visual design of the link, but I wanted to say that we might want to consider some sort of absolute positioning of it to avoid situations like the following where the inline content needs to reflow sometimes:

![Screen Recording 2022-09-02 at 1 33 46 PM](https://user-images.githubusercontent.com/212918/188231925-cd754ff6-ae7c-4a4d-9821-a7929ad4dffb.gif)

hqhhuang commented:
**URL capitalization:**
> This is because some come directly from docc (some have uppercase initials) and other we add them depending on the title (these ones are in lowercase). The anchor's names don't matter if they are in lowercase or not, as long as they point to an existing id with the same name.

What are some existing examples of ones coming directly from docc? I think if we opted for all lowercase for Docc-render generated anchors, docc should also provide all lower case anchors. Is there a particular reason some of the docc ones have upper case initials?

**Scroll jump:**
> This is because of the way we scroll to the section. We first scroll into the view where the title is, and then we scrollBy to overcome the navigation height. You can see it here:
> 
> https://github.com/apple/swift-docc-render/blob/f5c1b7fd271518503e85f7a5d2dd5f993ded17be/src/mixins/scrollToElement.js#L26-L27

Is there a difference between how Chrome and Safari handles the scroll? If this is the issue, I would sort of expect it to happen in both places. 

**Updating URL based on anchors:**
> Could you fill a radar or Github issue for it? Thanks!

rdar://99584817

dobromir-hristov commented on src/components/ContentNode/LinkableHeading.vue:
```diff
@@ -0,0 +1,90 @@
+  color: var(--colors-text, var(--color-text));
+  text-decoration: none;
+
+  // &:after {
```
👀

dobromir-hristov reviewed (commented):
I tested and it works OK. I think you havent missed any places where we use headings. 

Just make sure you fix what Marcus pointed out, with the jumping text on hover.

marinaaisa commented:
> I know you're still iterating on the visual design of the link, but I wanted to say that we might want to consider some sort of absolute positioning of it to avoid situations like the following where the inline content needs to reflow sometimes:

@mportiz08 oh true, fixed: https://github.com/apple/swift-docc-render/pull/408/commits/23fd3c74b803cf0a4909674154cb2b2280a090e6

but yes, the design specs aren't closed yet

marinaaisa commented:
@swift-ci test

marinaaisa commented:
@swift-ci test

marinaaisa commented:
@swift-ci test

marinaaisa commented:
@swift-ci test

mportiz08 commented on src/components/ContentNode/LinkableHeading.vue:
```diff
@@ -0,0 +1,84 @@
+$icon-margin: 7px;
+
+.header-anchor {
+  color: var(--colors-text, var(--color-text));
```
I think we should potentially remove this since these headings are currently being assigned the `--colors-header-text` value in a global/base h1-h6 rule, and this could potentially change the API, even though it's not noticeable at the moment since these colors are probably the same by default.

If removing it isn't sufficient, we should consider changing it to use `--colors-header-text`, but I would recommend just removing this line if possible to make it as close as possible to what it was before.

dobromir-hristov replied:
This should do it :) It will override the `color` for anchors and inherit the heading color.

```suggestion
  color: inherit;
```

marinaaisa replied:
I applied `color: inherit;`, good idea @dobromir-hristov !

marinaaisa replied:
and thanks @mportiz08 for pointing it out!

mportiz08 commented on src/styles/core/_vars.scss:
```diff
@@ -34,6 +34,9 @@ $article-stacked-margin-small: 20px;
 $topic-link-icon-spacing: 1rem;
 
+// Icons
+$icon-size: 16px;
```
Feel free to ignore this, but I think a name like `$icon-size-default` might be a little safer, because this one kind of implies that all icons will always be this size. That might be the case now (honestly not 100% sure), but I see that there could be a future situation where we have some icons of different sizes in different places in the UI and this variable name might be confusing then.

marinaaisa replied:
done!

mportiz08 reviewed (commented):
Awesome work @marinaaisa! This is working great, and I think it's almost ready to merge.

I just noticed one minor potential issue to look at with the color of the new linkable heading element.

dobromir-hristov reviewed (commented):
This looks good to me, just check the things that Marcus pointed out.

SamLanier commented:
New design approved.

marinaaisa commented:
@swift-ci test
